modules/command-engine/core/src/main/scala/surge/internal/kafka/KafkaProducerActorImpl.scala

```diff
@@ -225,18 +249,44 @@ class KafkaProducerActorImpl(
   private def handleProcessingProducerMessage(message: KafkaProducerActorImpl.KafkaProducerActorMessage, state: KafkaProducerActorState): Unit = {
     message match {
       case msg: Publish =>
-        context.become(processing(state.addPendingWrites(sender(), msg)))
+        val maybeTrackerWithExpiry = publishTrackerStateManager.tracker(msg.batchId, state)
+        maybeTrackerWithExpiry match {
+          case Some(trackerWithExpiry) =>
+            if (!trackerWithExpiry.tracker.dataWasPublished()) {
+              if (!trackerWithExpiry.expired) {
+                // Avoid duplicating a write that is already pending.
+                if (!publishTrackerStateManager.stateHasPendingWrites(msg.batchId, state)) {
+                  context.become(processing(state.addPendingWrites(sender(), msg)))
+                } else {
+                  context.become(processing(state))
```

Comment:
In this case (if I follow correctly) data hasn't been published but we're already tracking things - i.e. there is a current write to Kafka in progress. Wouldn't we want to update the original sender for the message so that the new asker is notified of completion rather than the old one (who has since given up)?

Reply:
done